aten/src/ATen/native/cuda/FractionalMaxPool2d.cu

```diff
@@ -0,0 +1,367 @@
+#include "ATen/ATen.h"
+#include "ATen/AccumulateType.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+#include "ATen/cuda/CUDAContext.h"
+#include "ATen/cuda/detail/IndexUtils.cuh"
+#include "ATen/cuda/detail/TensorInfo.cuh"
+#include "ATen/cuda/detail/KernelUtils.h"
+#include "ATen/NativeFunctions.h"
+#include "ATen/TensorUtils.h"
+#include "ATen/Utils.h"
+#include "c10/util/Exception.h"
+
+#include <algorithm>
+#include <cfloat>
+#include <cmath>
+
+namespace at {
+namespace native {
+
+using namespace at::cuda::detail;
+
+namespace {
+
+template <typename scalar_t, typename accscalar_t>
+__device__ inline int get_interval(accscalar_t sample,
+  int index, int inputSize, int outputSize, int poolSize) {
+  accscalar_t alpha = (accscalar_t)(inputSize - poolSize) / (accscalar_t) (outputSize - 1);
+  if (index == outputSize - 1) {
+    return inputSize - poolSize;
+  } else {
+    return (int) ((index + sample) * alpha) - (int) (sample * alpha);
+  }
+}
+
+template <typename scalar_t, typename index_t>
+__device__ inline scalar_t* get_ref_by_coord(
+  TensorInfo<scalar_t, index_t> tensor_info,
+  int ndims, index_t batch, index_t plane,
+  index_t h, index_t w) {
+  index_t offset = plane * tensor_info.strides[ndims - 3] +
+    h * tensor_info.strides[ndims - 2] + w * tensor_info.strides[ndims - 1];
+  if(ndims == 4) {
+    offset += batch * tensor_info.strides[0];
+  }
+  return tensor_info.data + offset;
+}
+
+template <typename scalar_t>
+__global__ void fractional_max_pool2d_out_cuda_frame(
+  TensorInfo<scalar_t, int> output,
+  TensorInfo<int64_t, int> indices,
+  TensorInfo<scalar_t, int> input,
+  TensorInfo<scalar_t, int> samples,
+  int poolSizeH, int poolSizeW,
+  int PoolSizeWStatic) {
```

Comment:
Hmm. Once this is detemplatized, does the loop unrolling still happen? I'm not sure if this function gets inlined to its call site.

Reply:
I don't know the answer for this one. I believe the behavior is the same as the TH implementation. Do you have some suggestion to test / fix?

Reply:
The change has performance implications, so there are two ways to check: (1) look at the assembly before and after and see if the loop is unrolled or not; (2) run benchmarks to see if it affected anything or not.

Reply:
I made an update to template on PoolSizeWStatic. This keeps fidelity to the TH implementation. There is a comment in TH implementation that claims this ensures unroll when de-templated. Benchmark test result seems also confirm this claim: 5.59916s without unroll v.s. 5.34821s with unroll.